File: problems/leetcode/0800/815/815.py
# 815. Bus Routes
# https://leetcode.com/problems/bus-routes/description/

from collections import defaultdict, deque
from typing import *

class Solution:
    def numBusesToDestination(self, routes: List[List[int]], source: int, target: int) -> int:
        stopNumToRouteIndices = defaultdict(list[int])
        for i, route in enumerate(routes):
            for stop in route:
                stopNumToRouteIndices[stop].append(i)

        q = deque()
        q.append((source, 0))

        visitedStopNums, visitedRouteIndices = set(), set()
        visitedStopNums.add(source)
        while q:
            stopNum, move = q.popleft()

            if stopNum == target:
                return move

            for routeIndex in stopNumToRouteIndices[stopNum]:
                if routeIndex in visitedRouteIndices:
                    continue

                visitedRouteIndices.add(routeIndex)
                for nextStopNum in routes[routeIndex]:
                    if nextStopNum in visitedStopNums:
                        continue

                    visitedStopNums.add(nextStopNum)
                    q.append((nextStopNum, move + 1))

        return -1

File: problems/leetcode/0800/815/815.cpp
// 815. Bus Routes
// https://leetcode.com/problems/bus-routes/

using namespace std;

#include <map>
#include <queue>
#include <set>

typedef pair<int, int> pii; // (stop, busCount)

class Solution {
public:
  int numBusesToDestination(vector<vector<int>> &routes, int source,
                            int target) {
    map<int, set<int>> availableBuses;

    for (int i = 0; i < routes.size(); i++) {
      for (auto &stop : routes[i]) {
        if (availableBuses.find(stop) == availableBuses.end()) {
          availableBuses[stop] = set<int>({i});
        } else {
          availableBuses[stop].insert(i);
        }
      }
    }

    queue<pii> q;
    q.push({source, 0});

    set<int> visitedStops;
    visitedStops.insert(source);

    while (!q.empty()) {
      pii stopAndBusCount = q.front();
      int stop = stopAndBusCount.first, busCount = stopAndBusCount.second;
      q.pop();

      if (stop == target) {
        return busCount;
      }

      for (auto &busIndex : availableBuses[stop]) {
        for (auto &availableStop : routes[busIndex]) {
          if (visitedStops.find(availableStop) == visitedStops.end()) {
            visitedStops.insert(availableStop);
            q.push({availableStop, busCount + 1});
          }
        }

        routes[busIndex].clear();
      }
    }

    return -1;
  }
};
